include/libsemigroups/ranges.hpp

```diff
@@ -38,28 +38,101 @@
+  //! ``libsemigroups`` that extend the functionality of range objects from
+  //! [rx::ranges][].
+  //!
+  //! Please consult the [rx::ranges][] [README][] file for more information
+  //! about ranges in general and the existing functionality of [rx::ranges][].
+  //!
+  //! Range objects are used extensively elsewhere in ``libsemigroups``, and
+  //! are returned by many functions.
+  //!
+  //! \note So that range classes in ``libsemigroups`` can be used efficiently
+  //! with the functionality of [rx::ranges][], the usual naming conventions in
+  //! ``libsemigroups`` are not used for the member functions of range objects.
+  //! In particular, none of the member functions required by [rx::ranges][]
+  //! check their arguments, but they do not have the suffix ``_no_checks``.
+  //!
+  //! [rx::ranges]: https://github.com/simonask/rx-ranges/
+  //! [README]: https://github.com/simonask/rx-ranges/blob/develop/README.md
+
   ////////////////////////////////////////////////////////////////////////
   // Custom ranges
   ////////////////////////////////////////////////////////////////////////
 
-  // TODO(doc)
-  struct random {
+  //! \ingroup ranges_group
+  //!
+  //! \brief Aggregator for choosing a single random item in a range.
+  //!
+  //! Defined in ``ranges.hpp``.
+  //!
+  //! This struct can be used to select a single random item from an input
+  //! range. This can be extremely inefficient if the input range is large,
+  //! because every item in the range prior to the randomly chosen one must be
+  //! iterated through when ``get`` is called.
+  //!
+  //! \par Example
+  //! \code
+  //! auto wg = to_word_graph(4, [[0, 1], [1, 0], [2, 2]]);
+  //! Paths p(wg);
+  //! p.source(0).max(10);
+  //! p.count();            // returns 1023
+  //! (p | Random()).get(); // returns random path in p
+  //! \endcode
+  // TODO(2) this should be able to emit any number of random items not only
+  // one.
+  struct Random {
+    //! \brief Default constructor.
+    //!
+    //! Default constructor.
+    Random() = default;
+
+    //! \brief Default copy constructor.
+    //!
+    //! Default copy constructor.
+    Random(Random const&) = default;
+
+    //! \brief Default move constructor.
+    //!
+    //! Default move constructor.
+    Random(Random&&) = default;
+
+    //! \brief Default copy assignment operator.
+    //!
+    //! Default copy assignment operator.
+    Random& operator=(Random const&) = default;
+
+    //! \brief Default move assignment operator.
+    //!
+    //! Default move assignment operator.
+    Random& operator=(Random&&) = default;
+
     template <typename InputRange>
     struct Range {
       using output_type = typename InputRange::output_type;
 
-      static constexpr bool is_finite     = true;
-      static constexpr bool is_idempotent = false;
+      // TODO(1) this is probably not correct, should depend on InputRange
+      static constexpr bool is_finite = true;
```

Comment:
```suggestion
      static constexpr bool is_finite = rx::is_finite_v<InputRange>;
```
Something like this?